sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala

```diff
@@ -40,23 +40,64 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
    * @param id buffer ID to associate with this buffer
    * @param table cudf table based from the contiguous buffer
    * @param contigBuffer device memory buffer backing the table
+   * @param tableMeta metadata describing the buffer layout
    * @param initialSpillPriority starting spill priority value for the buffer
    */
   def addTable(
       id: RapidsBufferId,
       table: Table,
       contigBuffer: DeviceMemoryBuffer,
+      tableMeta: TableMeta,
       initialSpillPriority: Long): Unit = {
-    val buffer = uncompressedBufferFromTable(id, table, contigBuffer, initialSpillPriority)
+    val buffer = new RapidsDeviceMemoryBuffer(
+      id,
+      contigBuffer.getLength,
+      tableMeta,
+      Some(table),
+      contigBuffer,
+      initialSpillPriority)
+    try {
+      logDebug(s"Adding table for: [id=$id, size=${buffer.size}, " +
+          s"meta_id=${buffer.meta.bufferMeta.id}, meta_size=${buffer.meta.bufferMeta.size}]")
+      addBuffer(buffer)
+    } catch {
+      case t: Throwable =>
+        buffer.free()
```

Comment:
why not `closeOnExecpt`?

Reply:
Because `close` and `free` are semantically different.  Calling `close` on a freshly created buffer will throw an "already closed" exception since it was never acquired to "open" it.  `free` will release the resources if the refcount is zero, and that's what we need to do in this case.